sycl/include/sycl/reduction.hpp

```diff
@@ -77,20 +77,24 @@ template <typename T, class BinaryOperation, int Dims, size_t Extent,
 class reducer;
 
 namespace detail {
+
+#ifdef SYCL_DETERMINISTIC_REDUCTION
+// Act as if all operators require determinism.
+template <typename T> struct IsDeterministicOperator : std::true_type {};
+#else
+// Each operator declares whether determinism is required.
+template <typename T> struct IsDeterministicOperator : std::false_type {};
+#endif
+
 // This type trait is used to detect if the atomic operation BinaryOperation
 // used with operands of the type T is available for using in reduction.
 // The order in which the atomic operations are performed may be arbitrary and
 // thus may cause different results from run to run even on the same elements
-// and on same device. The macro SYCL_REDUCTION_DETERMINISTIC prohibits using
-// atomic operations for reduction and helps to produce stable results.
-// SYCL_REDUCTION_DETERMINISTIC is a short term solution, which perhaps become
-// deprecated eventually and is replaced by a sycl property passed to reduction.
+// and on same device.
 template <typename T, class BinaryOperation>
 using IsReduOptForFastAtomicFetch =
-#ifdef SYCL_REDUCTION_DETERMINISTIC
-    std::bool_constant<false>;
-#else
-    std::bool_constant<((is_sgenfloat_v<T> && sizeof(T) == 4) ||
+    std::bool_constant<!IsDeterministicOperator<BinaryOperation>::value &&
```

Comment:
nit: I prefer we provide `inline constexpr bool` if a trait is used more than once.

Reply:
I agree this would be nice, but all the traits in the reduction implementation are defined as a `std::bool_constant` without a corresponding `inline constexpr bool`.  I'd prefer if somebody else could make that change later, to limit the scope of this PR.